Win ci ii

kwryankrattiger opened the pull request:
<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

Rebased version of old Windows CI branch.

Needs #42727 
Needs #39739

kwryankrattiger committed:
CI: Add an windows CI pipeline for testing ParaView

kwryankrattiger committed:
CI: Add an windows CI pipeline for testing ParaView

johnwparent committed:
Empty Commit to churn CI

johnwparent committed:
Empty Commit to churn CI

kwryankrattiger committed:
CI: Add PR triggers for windows CI

kwryankrattiger committed:
Win paths use \\

kwryankrattiger committed:
Remove quotes from windows script

kwryankrattiger committed:
Don't use SPACK_ROOT env var

kwryankrattiger committed:
Fixup paths for system software

kwryankrattiger committed:
Update some of the CI configs

kwryankrattiger committed:
Add compiler find and external find to generate

johnwparent committed:
Correct Spack powershell env script entrypoint

johnwparent committed:
Skip non Windows jobs

johnwparent committed:
Remove aws runs from Windows tests

johnwparent committed:
Proper pwsh env var setting

johnwparent committed:
Inherit base job

johnwparent committed:
check if spack env var is set

johnwparent committed:
Gitlab yaml did not like "env:"

johnwparent committed:
workaround powershell env shell limiation

johnwparent committed:
Quote python for shlex

johnwparent committed:
Externally detect non installable packages

johnwparent committed:
rm arch requirement

johnwparent committed:
Cleanup invalid yaml

johnwparent committed:
Fix more yaml errors

johnwparent committed:
Spack yaml scheme bizzareness

johnwparent committed:
install boto3

kwryankrattiger committed:
Disable non-windows stacks using bad tags

kwryankrattiger committed:
Disable non-windows stacks using bad tags

kwryankrattiger committed:
Fixup windows configs

johnwparent committed:
Empty commit to try to trigger github

johnwparent committed:
Update job requirement name

johnwparent committed:
Fixup ci yaml useage

johnwparent committed:
Override global `any-job`

johnwparent committed:
Proper config override syntax

johnwparent committed:
Use posix style paths on powershell CI

johnwparent committed:
Convert process_command to run with Python

johnwparent committed:
Convert process_command to run with Python

johnwparent committed:
Update comments

johnwparent committed:
Update comments

kwryankrattiger commented:
CC: @johnwparent This is more an updated reference to try to get a view on where things are at and what still need to be ported yet or re-written.

kwryankrattiger commented on var/spack/repos/builtin/packages/python/package.py:
```diff
@@ -836,6 +836,7 @@ def command(self):
             else:
                 path = os.path.join(self.prefix, "python{0}.exe".format(ver))
             if os.path.exists(path):
+                path = f"""\"{path}\""""
```
Not sure if this is still needed or not.

kwryankrattiger committed:
Various updates to the configs

kwryankrattiger commented on share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml:
```diff
@@ -156,10 +156,11 @@ default:
+      "${SPACK_CI_CONFIG_ROOT}/mirrors.yaml"
     - spack config add -f "${SPACK_CI_CONFIG_ROOT}/mirrors.yaml"
-    - mkdir -p "${CI_PROJECT_DIR}/jobs_scratch_dir"
+    - mkdir "${CI_PROJECT_DIR}/jobs_scratch_dir"
```
`-p` shouldn't be required, but should make sure for other platforms.

kwryankrattiger commented on share/spack/gitlab/cloud_pipelines/configs/win64/ci.yaml:
```diff
@@ -0,0 +1,16 @@
+      - md ${SPACK_ARTIFACTS_ROOT}/user_data
+      - spack --backtrace ci rebuild > >(tee ${SPACK_ARTIFACTS_ROOT}/user_data/pipeline_out.txt) 2> >(tee ${SPACK_ARTIFACTS_ROOT}/user_data/pipeline_err.txt >&2)
+  - signing-job-remove:
+      script:: no-signing-on-windows
```
This disables signing and it seems like so long as we don't have gpg keys available nothing should be signed.

TODO: add signing for windows binaries.

johnwparent replied:
gpg is available on the runner image, should we setup some keys? Or do you and I have not have correct permissions on the Spack keyserver? (do we even have a keyserver?)

kwryankrattiger replied:
If we have gpg I would be interested to see how that works. I don't know if we mount the existing intermediate key already for the protected windows runners or not, I am not sure what kind of nuance that requires.

kwryankrattiger committed:
Configuration file updates to be consistent with other platforms

Changed windows-vis package to VTK@8.2

kwryankrattiger committed:
Enable build_systems for testing Linux in parallel

kwryankrattiger committed:
Remove python path quoting patch

johnwparent committed:
style

johnwparent committed:
style

kwryankrattiger committed:
Use the default servercore image

johnwparent committed:
Cleanup process_command

johnwparent committed:
Cleanup process_command

johnwparent committed:
re-enable ci tests

kwryankrattiger committed:
Use setup-env.ps1 to configure spack

kwryankrattiger committed:
Use the customized image

mvandenburgh commented on share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml:
```diff
@@ -213,6 +220,13 @@ default:
+.generate-win64:
+  extends: [ ".generate-common" ]
+  before_script:
+    - . .\share\spack\spack-env.ps1
```
```suggestion
    - . .\share\spack\setup-env.ps1
```

mvandenburgh commented on share/spack/gitlab/cloud_pipelines/configs/win64/ci.yaml:
```diff
@@ -0,0 +1,16 @@
+      after_script::
+      - Write-Output "Done"
+      before_script::
+      - . .\share\spack\spack-env.ps1
```
```suggestion
      - . .\share\spack\setup-env.ps1
```

kwryankrattiger committed:
Move environment variables to `variables` section

kwryankrattiger committed:
Fix typo

kwryankrattiger committed:
Move config scopes to be includes

kwryankrattiger committed:
Remove more config scopes

kwryankrattiger committed:
Add scripts root to variables

mvandenburgh commented on share/spack/gitlab/cloud_pipelines/scripts/common/expand_vars.py:
```diff
@@ -0,0 +1,10 @@
+
+args = parser.parse_args()
+
+args.out.write(os.path.expandvars(args.intput.read()))
```
```suggestion
args.out.write(os.path.expandvars(args.input.read()))
```

mvandenburgh committed:
Escape backslashes in windows paths in yaml files

The yaml parser doesn't seem to like the unescaped backslashes here.

kwryankrattiger committed:
Fix typo

Co-authored-by: Mike VanDenburgh <37340715+mvandenburgh@users.noreply.github.com>

kwryankrattiger committed:
Extend generate-windows with base-job

mvandenburgh committed:
spack env: Allow `C` "protocol" for config_path

When running spack on windows, a path beginning with `C://...` is a valid path.

kwryankrattiger committed:
Schema fixup

mvandenburgh committed:
Convert `tee` command to powershell equivalent

kwryankrattiger committed:
More setup for windows build-job

kwryankrattiger committed:
!fixup Don't prune untouched for testing

kwryankrattiger committed:
Remove makefile from ci rebuild

kwryankrattiger committed:
GPG use llnl.util.filesystem.getuid

kwryankrattiger committed:
Remove unused lines

kwryankrattiger commented on lib/spack/spack/cmd/ci.py:
```diff
@@ -38,6 +38,10 @@ def deindent(desc):
 
 
+def encode_path(path: str) -> str:
+    return path.encode("unicode-escape".decode())
```
@johnwparent Am I missing something about these lines from the cherry-pick?

```
  File "C:\builds\spack\spack\lib\spack\spack\cmd\ci.py", line 584, in ci_rebuild
    [SPACK_COMMAND, "-e", encode_path(env.path), "bootstrap", "now"],
                          ^^^^^^^^^^^^^^^^^^^^^
  File "C:\builds\spack\spack\lib\spack\spack\cmd\ci.py", line 42, in encode_path
    return path.encode("unicode-escape".decode())
                       ^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'str' object has no attribute 'decode'. Did you mean: 'encode'?
```

johnwparent replied:
~Yeah I missed the commit where I patch that typo, should be 8ff51a4fc50405be31b7b8bc6834f82a3ca16620~

You made the same change before I could finish commenting, disregard.

kwryankrattiger committed:
Fix tyop in encode_path

johnwparent committed:
patch mistake

tgamblin committed:
bugfix: add arg to `write_line_break()` in `spack_yaml`

`ruamel`'s `Emitter.write_line_break()` method takes an extra argument that we forgot to
implement in our custom emitter.

johnwparent committed:
Update cmd ci for Windows

johnwparent committed:
Cleanup style

johnwparent committed:
rm debug statements

johnwparent committed:
process_command modify executor

johnwparent committed:
rm dnagling ref

kwryankrattiger committed:
Double quote arguments

johnwparent committed:
Don't fail on error

kwryankrattiger committed:
replace " with '

kwryankrattiger committed:
Pass cdash args with =

kwryankrattiger committed:
Escape parens in CMD script

kwryankrattiger committed:
only build bzip for  testing

kwryankrattiger committed:
Merge remote-tracking branch 'jp/ci/refactor-reproduce-to-produce-python' into win_ci_ii

kwryankrattiger committed:
escape parens doesn't only apply to paths

kwryankrattiger commented:
Pulled in the latest script updates from #39739
Added escaping for `(` and `)` on windows
fixup: swap vtk for bzip2

kwryankrattiger committed:
Merge remote-tracking branch 'origin/bugfix-spack-yaml-windows-line-break' into win_ci_ii

kwryankrattiger committed:
Install deps

kwryankrattiger committed:
Call spack using ps1 script wrapper

kwryankrattiger committed:
drop audit check

kwryankrattiger commented:
This work has been completed in another set of PRs.
